README.md
![alt text for screen readers](https://raw.githubusercontent.com/Siumauricio/rippleui/main/docs/public/logo.png "RippleUI Logo")

<h1 align="center">Ripple UI</h1>

<a href="https://www.npmjs.com/package/rippleui">
    <img src="https://img.shields.io/npm/dm/rippleui.svg?style=flat-round" alt="npm downloads">
  </a>
<a href="https://www.npmjs.com/package/rippleui">
    <img src="https://img.shields.io/npm/v/rippleui?color=green" alt="npm downloads">
  </a>
  <a href="https://www.npmjs.com/package/rippleui">
    <img src="https://img.shields.io/npm/l/rippleui" alt="npm downloads">
  </a>

<a href="https://www.npmjs.com/package/rippleui"><img src="https://badgen.net/github/stars/Siumauricio/rippleui" ></a>

![GitHub last commit (by committer)](https://img.shields.io/github/last-commit/siumauricio/rippleui)

Welcome to the RippleUI repository! This is the home for all of the components, guidelines, and tools that make up our design system.

## Documentation

Visit https://ripple-ui.com to view the full documentation.

## 🚀 Usage

To use the RippleUI components, you will need to install the package from npm.

```bash
npm install rippleui
```

Once you have installed the package, you can import the plugin into your `tailwind.config.js`.

```js
module.exports = {
  // ... the rest of your config
  plugins: [require("rippleui")],
};
```

To use a component, simply include the appropriate HTML markup in your project and apply the relevant classes using the class naming.

```html
<button class="btn-primary btn">Click me!</button>
```

## 🎨 Components

Here is a list of all of the components that are currently available in RippleUI.

- Button
- Input
- Select
- Checkbox
- Radio
- Switch
- Textarea
- Modal
- Button-Group
- Tooltip
- Dropdown
- Badge
- Navbar
- Avatar
- Breadcrumb
- Drawer
- Alert
- Table
- Card
- Accordion
- Tabs
- Pagination
- Divider
- Skeleton
- Stepper
- Link
- Dots (Status)
- Range
- Progress
- KBD
- Forms
- Menu
- Popover
- Sidebar
- Spinners / Loaders
- Input File
- More components soon....

## 🤝 Contributing

We welcome contributions to the design system! If you have ideas for new components, improvements to existing ones, or updates to the guidelines and tools, please open a pull request with your proposed changes.

## License

RippleUI is licensed under the [MIT License](LICENSE).


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

### [1.12.1](https://github.com/Siumauricio/rippleui/compare/v1.12.0...v1.12.1) (2023-04-23)


### Bug Fixes

* **file-input:** add support for firefox and add rounded to button on safari ([#95](https://github.com/Siumauricio/rippleui/issues/95)) ([14a3262](https://github.com/Siumauricio/rippleui/commit/14a3262d1ec44b3e57b4cf6d3c4d4c7b457c0c20))

## [1.12.0](https://github.com/Siumauricio/rippleui/compare/v1.11.0...v1.12.0) (2023-04-16)


### Features

* **input-file:** input-file component ([d187acf](https://github.com/Siumauricio/rippleui/commit/d187acfea2e9dbd0574474dd73860a1f7eaeb201))
* **v1.11.0:** docs ([557ce78](https://github.com/Siumauricio/rippleui/commit/557ce78963baf2d4328d4a93a9d53398c7703669))

## [1.11.0](https://github.com/Siumauricio/rippleui/compare/v1.10.1...v1.11.0) (2023-04-09)

### Features

- **spinner:** spinner component ([eb83aff](https://github.com/Siumauricio/rippleui/commit/eb83aff381ee98f99dc15bfaa6bcd098c1c65960))

### Refactor

- **switch:** set gray-7 to border hover ([4f5325b](https://github.com/Siumauricio/rippleui/commit/4f5325b7d80f4959aaa4e6e4caeabec1de769471))

### [1.10.1](https://github.com/Siumauricio/rippleui/compare/v1.10.0...v1.10.1) (2023-03-30)

### Refactor

- remove default font ([4b8b4a9](https://github.com/Siumauricio/rippleui/commit/4b8b4a914a6aafae2d5f8cfa2cee7f3ab1354d45))

## [1.10.0](https://github.com/Siumauricio/rippleui/compare/v1.9.2...v1.10.0) (2023-03-20)

### Features

- **button, checkbox, input, radio, select, textarea:** add solid variant ([53a0e54](https://github.com/Siumauricio/rippleui/commit/53a0e546389d0518bc969250eea6e5bdda5261d0))
- **sidebar:** sidebar component ([e99276f](https://github.com/Siumauricio/rippleui/commit/e99276f541236cd53a589bf1bfcb40c559c61aee))

### Bug Fixes

- **tooltip:** set value instead variable ([28c3aa9](https://github.com/Siumauricio/rippleui/commit/28c3aa9a837d10500618429960edf392493da43e))

### [1.9.2](https://github.com/Siumauricio/rippleui/compare/v1.9.1...v1.9.2) (2023-03-05)

### Refactor

- **checkbox, radio, range, switch:** set static white value ([f3ccb18](https://github.com/Siumauricio/rippleui/commit/f3ccb1859fbb7f9e8a5afec9976620566584f776)) [#39](https://github.com/Siumauricio/rippleui/issues/39)
- **config:** remove white and black colors ([ad88491](https://github.com/Siumauricio/rippleui/commit/ad884915099bb6be0137cfa0c4de864a0f61a8f0)) [#39](https://github.com/Siumauricio/rippleui/issues/39)

### [1.9.1](https://github.com/Siumauricio/rippleui/compare/v1.9.0...v1.9.1) (2023-03-04)

### Features

- **config:** add standalone-cli executables ([f170e2b](https://github.com/Siumauricio/rippleui/commit/f170e2b4f49f3501be10d182dc87ec1038e5419b))
- **skeleton:** add static variant ([c7e8137](https://github.com/Siumauricio/rippleui/commit/c7e813704a1fadfc809b675cc65625f7cdc50f9a))

### Refactor

- **config:** add example of standalone cli ([2019a8b](https://github.com/Siumauricio/rippleui/commit/2019a8bfc7a96d338fcec174c1544120462353d4))
- **menu,popover,button,dropdown:** prevent components vibrate on click ([6b41ef8](https://github.com/Siumauricio/rippleui/commit/6b41ef8bca9cc6b004993c13f76491e59be0fd78))

### Breaking Changes

- Change tokens colors to not clash with tailwind colors ([c2f1171](https://github.com/Siumauricio/rippleui/commit/c2f1171302dd7aae119a8b97fdaec2441c960c4e)) #39
  Since the classes that we use are clashing with tailwindcss default colors.
  The colors have the same name the only thing will change is the last `token` and follow the same semantic as [RadixUI](https://www.radix-ui.com/docs/colors/palette-composition/the-scales#colors) e.g:

|  Old Token  | New Token |
| :---------: | :-------: |
| bg-blue-500 | bg-blue-6 |
| bg-red-900  | bg-red-10 |

## [1.9.0](https://github.com/Siumauricio/rippleui/compare/v1.8.1...v1.9.0) (2023-02-12)

### Features

- **menu:** menu component ([f151b47](https://github.com/Siumauricio/rippleui/commit/f151b470487e32a9d324b13dd4aedf4605776163))
- **popover:** popover component ([3ff80c2](https://github.com/Siumauricio/rippleui/commit/3ff80c2d5374f49e80fe99c107d57db507024151))

### Refactor

- **button:** set text color to default variant ([895d034](https://github.com/Siumauricio/rippleui/commit/895d034861083041047358db14968cc0942d2aca))
- **dropdown:** add center bottom and top variant ([4793894](https://github.com/Siumauricio/rippleui/commit/4793894837884ac3e472301e91d85f51b9c33793))
- **dropdown:** removed border and add scale to dropdown item ([2dacaf7](https://github.com/Siumauricio/rippleui/commit/2dacaf7b807fe67a1e5a18d3cbfdf772a7238924))
- **dropdown:** removed margin to menu ([20c7502](https://github.com/Siumauricio/rippleui/commit/20c7502375ce3e46d77802c59a47b419b6b05ca2))
- **dropdown:** set width only to a main menu class ([83a7417](https://github.com/Siumauricio/rippleui/commit/83a7417c4f952e029852b2dd6047f9370e62b4b2))
- **dropdown:** simplify selectors and change transition duration on transform ([7e31aa6](https://github.com/Siumauricio/rippleui/commit/7e31aa6394aa7ea02a5352ee433873ae17e98a89))

### [1.8.1](https://github.com/Siumauricio/rippleui/compare/v1.8.0...v1.8.1) (2023-02-03)

### Features

- **forms:** forms component ([dbea29a](https://github.com/Siumauricio/rippleui/commit/dbea29a991461ba9b2b3a8ad8cb93a349a87deaf))

### Bug Fixes

- **progress:** Fix indeterminate progress bar on cross-platform ([92337b6](https://github.com/Siumauricio/rippleui/commit/92337b67a6b3d38b0aa725f5890e631d07a81fe9))

### Refactor

- **dot:** set display value ([632a425](https://github.com/Siumauricio/rippleui/commit/632a42583adb77f8220459942aceff23cdf42803))
- **progress:** set width by default ([31c8a16](https://github.com/Siumauricio/rippleui/commit/31c8a1601ddd969531553d0fee9cbeec64f715e4))

config/types.ts
import { applyOpacityValue } from "./utils/applyOpacityValue";
export type Colors =
  | "blue"
  | "green"
  | "yellow"
  | "red"
  | "cyan"
  | "pink"
  | "purple"
  | "gray";

type PalleteScale =
  | "50"
  | "100"
  | "200"
  | "300"
  | "400"
  | "500"
  | "600"
  | "700"
  | "800"
  | "900";

type BaseScale =
  | "primary"
  | "secondary"
  | "success"
  | "error"
  | "warning"
  | "neutral";

type BackgroundScale =
  | "backgroundPrimary"
  | "backgroundSecondary"
  | "border"
  | "black"
  | "white";

type TextScale =
  | "content1"
  | "content2"
  | "content3"
  | "whiteInverted"
  | "blackInverted";

type DefaultSchema = "background-color" | "color";

type BackgroundVariablesScale = `--${BackgroundScale}`;

type BaseVariablesScale = `--${BaseScale}`;

type TextVariablesScale = `--${TextScale}`;

type PalleteVariablesScale<T extends Colors> = `--${T}-${PalleteScale}`;

type PalleteVariablesAll = `--${Colors}-${PalleteScale}`;

type OpacityFunc = ReturnType<typeof applyOpacityValue>;

export type DefaultSchemaColors = {
  [key in DefaultSchema]: string;
};

export type PalleteColors = {
  [key in PalleteScale]: OpacityFunc;
};

export type BackgroundColors = {
  [key in BackgroundScale]: OpacityFunc;
};

export type TextColors = {
  [key in TextScale]: OpacityFunc;
};

export type BaseColors = {
  [key in BaseScale]: OpacityFunc;
};

export type Pallete = {
  [key in Colors]: PalleteColors;
};

export type Theme = {
  colors: BackgroundColors & TextColors & BaseColors & Pallete;
};

export type BackgroundVariables = {
  [key in BackgroundVariablesScale]: string;
};

export type TextVariables = {
  [key in TextVariablesScale]: string;
};

export type BaseVariables = {
  [key in BaseVariablesScale]: string;
};

export type PalleteVariables<T extends Colors> = {
  [key in PalleteVariablesScale<T>]: string;
};

// export type Variables = {
//   [key in Colors]: PalleteVariables<key>;
// };

export type Variables = {
  [key in PalleteVariablesAll]: string;
};

type RecursivePartial<T> = {
  [P in keyof T]?: RecursivePartial<T[P]>;
};

export type ThemeVariables = {
  colors: {
    [key in BackgroundVariablesScale]: string;
  } & {
    [key in BaseVariablesScale]: string;
  } & {
    [key in TextVariablesScale]: string;
  } & {
    [key in PalleteVariablesAll]: string;
  };
};

export type ThemeType = RecursivePartial<ThemeVariables>;
